File: s2_benchmarks/cyclist/schorr_waite/schorr_waite.c
#include <stdlib.h>
#include <stdbool.h>

struct node {
  bool m;
  bool c;
  struct node* l;
  struct node* r;
};

void schorr_waite(struct node* root)
{
  //pre
  struct node* t = root;
  struct node* p = 0;
  while(p != 0 || (t != 0 && ! (t->m)))
  {
    //loop
    if(t == 0 || t->m) {
      if(p->c) {
        struct node* q = t;
        t = p;
        p = p->r;
        t->r = q;
      } else {
        struct node* q = t;
        t = p->r;
        p->r = p->l;
        p->l = q;
        p->c = true;
      }
    } else {
      struct node* q = p;
      p = t;
      t = t->l;
      p->l = q;
      p->m = true;
      p->c = false;
    }
  }
  //post
}

struct node * makeNode()
{
    //pre
	struct node * n = malloc(sizeof(struct node));
	n->m = false;
	n->c = false;
	n->l = 0;
	n-> r = 0;
    //post
	return n;
}

File: python/schorr_waite.py
import random
import pprint

class node:
	def __init__(self, m = False , c = False, l = None, r = None):
		self.m = m
		self.c = c
		self.l = l
		self.r = r

def schorr_waite( root):
	t = root
	p = 0
	while p != 0 or (t != 0 and not (t.m))):
		if(t == 0 or t.m):
			if(p.c):
				q = t
				t = p
				p = p.r
				t.r = q
			else:
				q = t
				t = p.r
				p.r = p.l
				p.l = q
				p.c = True
		else:
			q = p
			p = t
			t = t.l
			p.l = q
			p.m = True
			p.c = False

def makeNode():
	n = node(False, False, None, None)
	return n

def main():
	root = makeNode()
	for i in range(1, 10):
		n = root
		p = None
		j = 0
		while n is not None:
			j = random.randint(-5, 5)
			p = n
			if j > 0:
				n = n.l
			else
				n = n.r
		if j > 0:
			p.l = makeNode()
		else:
			p.r = makeNode()
		i = i + 1
	schorr_waite(root)

if __name__=="__main__":
	main()
